### PR TITLE
Support for test run labels/selectors

### LINKED ISSUE #379: Conditional execution of test steps
**What would you like to be added**:

I would like to run a test step only if a condition is met. This would allow me to install and test cluster specific resources.

For example, Helm has a way to check if a certain API is available in the cluster by querying the `capabilities` object and install custom resources of the corresponding `kind`.

Here is an snippet  that creates a SCC resource only when the `security.openshift.io/v1` API is available:

```
{{ if .Capabilities.APIVersions.Has "security.openshift.io/v1" }}
---
apiVersion: security.openshift.io/v1
kind: SecurityContextConstraints
metadata:
  name: airflow-scc
...
{{ end }}
```

As you'd guess, the custom resource above is only created if the target cluster of the Helm chart is an OpenShift cluster.

In a similar fashion, I would like to be able to run a `kuttl` test step that creates  the same SCC only if the tests are running against OpenShift or to be more general, only if the API `security.openshift.io/v1` is available.

**Why is this needed**:

We write many tests and would like them to be reusable without modification across kubernetes implementations.

Currently, to run the same tests on KinD and OpenShift for example, we'd have to maintain two largely redundant code bases for each target cluster. This duplicates the effort to create and maintain these tests.

### PR DESCRIPTION
**Pull request by porridge**
# **What this PR does / why we need it**:

Make it possible to associate label sets with test runs and put label selectors for individual files [as discussed on Slack](https://kubernetes.slack.com/archives/CG3HTFCMV/p1687170514079399). The purpose is conditional ignoring of selected files with the goal of having a single test directory but with some variation to allow for testing on different platforms.

# Why a new resource type?

I was not able to find any appropriate place.
* A `TestStep` object applies to all files in a given step, and only single such object [is permitted](https://github.com/kudobuilder/kuttl/blob/f6d64c915c8dd9e2da354562c3d5c6fcf88aec2b/pkg/test/step.go#L527) per step. This might be too coarse granularity, since for example in a given step we might only want to vary some assertions and keep the apply files constant.
* Similarly, only a single `TestAssert` is used in a given step (although unlike for `TestStep`, additional `TestAssert` objects seem to be silently ignored rather than complained about). Also, `TestAssert` objects are only active if they appear in an NN-assert*.yaml file - they seem to be ignored in all other files, [including errors](https://github.com/kudobuilder/kuttl/issues/456) files and varying the apply files is a goal.

# Simple usage example

## Setup

```
[mowsiany@mowsiany x]$ head -n 99 tt/*
==> tt/01-assert-cat.yaml <==
apiVersion: kuttl.dev/v1beta1
kind: TestFile
testRunSelector:
  matchLabels:
    flavor: cat
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data:
  key: fluffy

==> tt/01-assert-dog.yaml <==
apiVersion: kuttl.dev/v1beta1
kind: TestFile
testRunSelector:
  matchLabels:
    flavor: dog
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data:
  key: rex

==> tt/01-create.yaml <==
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data:
  key: fluffy
[mowsiany@mowsiany x]$ make -C .. cli
make: Entering directory '/home/mowsiany/go/src/github.com/kudobuilder/kuttl'
go build -ldflags "-X github.com/kudobuilder/kuttl/pkg/version.gitVersion=0.15.0 -X github.com/kudobuilder/kuttl/pkg/version.gitCommit=e7d8400e -X github.com/kudobuilder/kuttl/pkg/version.buildDate=2023-08-22T08:49:58Z" -o bin/kubectl-kuttl ./cmd/kubectl-kuttl
make: Leaving directory '/home/mowsiany/go/src/github.com/kudobuilder/kuttl'
```

## A `flavor=cat` test run

```
[mowsiany@mowsiany x]$ ../bin/kubectl-kuttl test . --start-kind --timeout 15 --test-run-labels flavor=cat
2023/08/22 10:39:41 running without a 'kuttl-test.yaml' configuration
2023/08/22 10:39:41 kutt-test config testdirs is overridden with args: [ . ]
=== RUN   kuttl
    harness.go:465: starting setup
    harness.go:252: running tests with KIND.
    harness.go:176: temp folder created /tmp/nix-shell.mdjyec/nix-shell.UvjEuW/kuttl3988091570
    harness.go:158: Starting KIND cluster
    kind.go:66: Adding Containers to KIND...
    harness.go:278: Successful connection to cluster at: https://127.0.0.1:40829
    harness.go:363: running tests
    harness.go:75: going to run test suite with timeout of 15 seconds for each step
    harness.go:375: testsuite: . has 1 tests
=== RUN   kuttl/harness
=== RUN   kuttl/harness/tt
=== PAUSE kuttl/harness/tt
=== CONT  kuttl/harness/tt
Skipping file "/home/mowsiany/go/src/github.com/kudobuilder/kuttl/x/tt/01-assert-dog.yaml", label selector does not match test run labels.
    logger.go:42: 10:40:10 | tt | Creating namespace: kuttl-test-shining-cat
    logger.go:42: 10:40:10 | tt/1-create | starting test step 1-create
    logger.go:42: 10:40:10 | tt/1-create | ConfigMap:kuttl-test-shining-cat/test created
    logger.go:42: 10:40:10 | tt/1-create | test step completed 1-create
    logger.go:42: 10:40:10 | tt | tt events from ns kuttl-test-shining-cat:
    logger.go:42: 10:40:10 | tt | Deleting namespace: kuttl-test-shining-cat
=== NAME  kuttl
    harness.go:408: run tests finished
    harness.go:516: cleaning up
    harness.go:525: collecting cluster logs to kind-logs-1692693615
    harness.go:573: removing temp folder: "/tmp/nix-shell.mdjyec/nix-shell.UvjEuW/kuttl3988091570"
    harness.go:579: tearing down kind cluster
--- PASS: kuttl (36.29s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/tt (5.20s)
PASS
```

## A `flavor=dog` run

```
[mowsiany@mowsiany x]$ ../bin/kubectl-kuttl test . --start-kind --timeout 15 --test-run-labels flavor=dog
2023/08/22 10:40:33 running without a 'kuttl-test.yaml' configuration
2023/08/22 10:40:33 kutt-test config testdirs is overridden with args: [ . ]
=== RUN   kuttl
    harness.go:465: starting setup
    harness.go:252: running tests with KIND.
    harness.go:176: temp folder created /tmp/nix-shell.mdjyec/nix-shell.UvjEuW/kuttl3909311779
    harness.go:158: Starting KIND cluster
    kind.go:66: Adding Containers to KIND...
    harness.go:278: Successful connection to cluster at: https://127.0.0.1:40901
    harness.go:363: running tests
    harness.go:75: going to run test suite with timeout of 15 seconds for each step
    harness.go:375: testsuite: . has 1 tests
=== RUN   kuttl/harness
=== RUN   kuttl/harness/tt
=== PAUSE kuttl/harness/tt
=== CONT  kuttl/harness/tt
Skipping file "/home/mowsiany/go/src/github.com/kudobuilder/kuttl/x/tt/01-assert-cat.yaml", label selector does not match test run labels.
    logger.go:42: 10:41:01 | tt | Creating namespace: kuttl-test-integral-porpoise
    logger.go:42: 10:41:01 | tt/1-create | starting test step 1-create
    logger.go:42: 10:41:01 | tt/1-create | ConfigMap:kuttl-test-integral-porpoise/test created
    logger.go:42: 10:41:17 | tt/1-create | test step failed 1-create
    case.go:366: failed in step 1-create
    case.go:368: --- ConfigMap:kuttl-test-integral-porpoise/test
        +++ ConfigMap:kuttl-test-integral-porpoise/test
        @@ -1,8 +1,18 @@
         apiVersion: v1
         data:
        -  key: rex
        +  key: fluffy
         kind: ConfigMap
         metadata:
        +  managedFields:
        +  - apiVersion: v1
        +    fieldsType: FieldsV1
        +    fieldsV1:
        +      f:data:
        +        .: {}
        +        f:key: {}
        +    manager: kubectl-kuttl
        +    operation: Update
        +    time: "2023-08-22T08:41:01Z"
           name: test
           namespace: kuttl-test-integral-porpoise
         
        
    case.go:368: resource ConfigMap:kuttl-test-integral-porpoise/test: .data.key: value mismatch, expected: rex != actual: fluffy
    logger.go:42: 10:41:17 | tt | tt events from ns kuttl-test-integral-porpoise:
    logger.go:42: 10:41:17 | tt | Deleting namespace: kuttl-test-integral-porpoise
=== NAME  kuttl
    harness.go:408: run tests finished
    harness.go:516: cleaning up
    harness.go:525: collecting cluster logs to kind-logs-1692693682
    harness.go:573: removing temp folder: "/tmp/nix-shell.mdjyec/nix-shell.UvjEuW/kuttl3909311779"
    harness.go:579: tearing down kind cluster
--- FAIL: kuttl (49.76s)
    --- FAIL: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/kind-logs-1692693615 (5.20s)
        --- FAIL: kuttl/harness/tt (20.57s)
FAIL
[mowsiany@mowsiany x]$ 
```

# Real-world usage example

See [this test directory](https://github.com/stackrox/stackrox/tree/porridge/ROX-16724/operator/tests/central/basic-central) (see the files ending in `(non-)openshift.yaml`).

Fixes #379

### CODE REVIEW

**Review by kensipe — COMMENTED** [review]
nothing really blocking on this... it would be great to have a conversation around comments but if you feel good about.   Documentation is missing and we should create an issue to capture that.   we have a number of ways to skip or isolate tests down... like using `--test`.  I didn't think through it deeply... I suppose we can have awkward scenarios which is the responsibility of the user to understand.

something to be cautious of (going off intuition here).  It seems likely that we want to "wait" for a running object with a label... or have an object started outside of kuttl that has a label.  It seems we don't want to confuse users with different label context.   This one is a selector for a test label (which isn't an object running in the cluster... it is a set of declarations to use during a test).  Another type of selector is to select on running objects and I wonder if we need to delineate that... perhaps premature.   However I've been thinking a while now about adding kuttl labels or annotations on objects it creates.  We currently do not cleanup tests if we do not delete the namespace.   I was thinking we could do clean up if we labeled or annotated the kuttl created objects.   Great addition and thanks @porridge !

File: pkg/kuttlctl/cmd/values.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package cmd
```
[the remark this answers is not in the document]
Comment (porridge) [reply pkg/kuttlctl/cmd/values.go:1]: I called it values because the type implements a`pflag.Value` [interface](https://pkg.go.dev/github.com/spf13/pflag@v1.0.5#Value) and I thought it might be a good name if we add more types that can be used via that interface. Maybe I should have inserted a comment about this...

But sure, this can be `labels.go` too if we go for one file per type. Let me know which one you prefer.

File: pkg/apis/testharness/v1beta1/test_types.go [hunk 2]
```diff
@@ -13,6 +13,19 @@ type RestConfig struct {
...
 
+// TestFile contains attributes of a single test file.
+type TestFile struct {
+	// The type meta object, should always be a GVK of kuttl.dev/v1beta1/TestFile.
+	metav1.TypeMeta `json:",inline"`
+	// Set labels or the test suite name.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Which test runs should this file be used in. Empty selector matches all test runs.
+	TestRunSelector *metav1.LabelSelector `json:"testRunSelector,omitempty"`
```
[the remark this answers is not in the document]
Comment (porridge) [reply pkg/apis/testharness/v1beta1/test_types.go:24]: This particular arrangement of which side labels vs selectors go was @eddycharly 's idea, so maybe he can chime in on where it came from. Maybe it's following some pattern seen elsewhere?

I guess you can think about this in two ways:
1. (_this implementation_): a "test run" (i.e. a `kuttl` invocation) comes with a set of labels attached. These labels represent some characteristics of this test invocation. In my use case, there would would be a single label with just two possible values:`openshift=true` or `openshift=false`, but you could imagine multiple labels and values for a more elaborate test matrix (such as the cartesian product of `plaform={openshift,vanilla,eks},k8s-version={1.28,1.29,1.30}`). Then individual files in the test suite provide an optional label selector, which needs to match the label set in order for the file to be considered for the test run. Obviously the selector can contain a subset of label keys that the test run has. For example a given file might only select one or more `k8s-version` values and be neutral w.r.t. `platform`. The notable edge case is a file that (be default) has no selector attached - an empty selector matches all possible label sets, meaning this feature is fully backwards-compatible - a file without a `TestFile` is included in all test invocations, just like before this change.
2. (I haven't thought about this case earlier so I'm inventing things as I go.) Each _test file_ comes with an optional _label set_ (empty by default). Let's say that some file tests some API introduced in version 1.29 but also present in 1.30. Since a label set can only specify a single value for a given label key, we'd either have to allow multiple label sets per file (eek!) or introduce a copy of the file per k8s version (ouch) or instead of labeling by k8s version, label by feature availability `feature-foo={true,false}`. Then, each _test run_ can have a _label selector_ attached, and this selector would need to select the APIs supported. This could quickly create an explosion of labels and combinations. What's worse, files which are agnostic about a given API would have to somehow advertise that fact, in order to be selected by a selector which does mention a given API.

So it looks _to me_ that only solution 1 is in fact realistic.

**Review by kensipe — APPROVED** [review]
/lgtm

after a hangout conversation around this... this makes a lot of sense and it looks ready to go.
